pkg/controller/cloudprivateipconfig/cloudprivateipconfig_controller.go

```diff
@@ -0,0 +1,437 @@
+		ctx, cancel := context.WithTimeout(c.ctx, controller.ClientTimeout)
+		defer cancel()
+
+		// Get the latest item from the store before updating. We risk updating
+		// and overriding any client changes if we don't get the latest state of
+		// the object and update the status on that item. Ex: if we are in the
+		// middle of updating the status on item v1, after an update to the
+		// object v2 has been performed we'll override the changes made to v2
+		// back to v1 when updating below (because cloudPrivateIPConfig above
+		// will have the old state before any potential client changes).
+		cloudPrivateIPConfigLatest, err := c.cloudPrivateIPConfigLister.Get(cloudPrivateIPConfig.Name)
+		if err != nil {
+			klog.Infof("CloudPrivateIPConfig: %q was deleted while processing an status update", cloudPrivateIPConfig.Name)
+			return nil
+		}
+		cloudPrivateIPConfigLatest.Status = *status
+		updatedCloudPrivateIPConfig, err = c.cloudNetworkClient.CloudV1().CloudPrivateIPConfigs().UpdateStatus(ctx, cloudPrivateIPConfigLatest, metav1.UpdateOptions{})
+		return err
+	})
+	return updatedCloudPrivateIPConfig, err
+}
+
+// updateCloudPrivateIPConfig copies and updates the provided object and returns
+// the new object. The return value can be useful for recursive updates
+func (c *CloudPrivateIPConfigController) updateCloudPrivateIPConfigFinalizer(cloudPrivateIPConfig *cloudnetworkv1.CloudPrivateIPConfig) (*cloudnetworkv1.CloudPrivateIPConfig, error) {
+	updatedCloudPrivateIPConfig := &cloudnetworkv1.CloudPrivateIPConfig{}
+	err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		ctx, cancel := context.WithTimeout(c.ctx, controller.ClientTimeout)
+		defer cancel()
+
+		// See: updateCloudPrivateIPConfigStatus
+		cloudPrivateIPConfigLatest, err := c.cloudPrivateIPConfigLister.Get(cloudPrivateIPConfig.Name)
+		if err != nil {
+			klog.Infof("CloudPrivateIPConfig: %q was deleted while processing an update", cloudPrivateIPConfig.Name)
+			return nil
+		}
+		cloudPrivateIPConfigLatest.Finalizers = cloudPrivateIPConfig.Finalizers
+		updatedCloudPrivateIPConfig, err = c.cloudNetworkClient.CloudV1().CloudPrivateIPConfigs().Update(ctx, cloudPrivateIPConfigLatest, metav1.UpdateOptions{})
+		return err
+	})
+	return updatedCloudPrivateIPConfig, err
+}
+
+// computeOp decides on what needs to be done given the state of the object.
+func (c *CloudPrivateIPConfigController) computeOp(cloudPrivateIPConfig *cloudnetworkv1.CloudPrivateIPConfig) (string, string) {
+	// Delete if the deletion timestamp is set and we still have our finalizer listed
+	if !cloudPrivateIPConfig.DeletionTimestamp.IsZero() && controllerutil.ContainsFinalizer(cloudPrivateIPConfig, cloudPrivateIPConfigFinalizer) {
+		return "", cloudPrivateIPConfig.Status.Node
+	}
+	// Update if status and spec are different
+	if cloudPrivateIPConfig.Spec.Node != cloudPrivateIPConfig.Status.Node && cloudPrivateIPConfig.Status.Node != "" {
+		return "", cloudPrivateIPConfig.Status.Node
+	}
```

Comment:
> What tells us to re-add it in the next sync?

When the current status assignment finally and successfully is deleted: we'll finish by updating the status in the delete code to `""`. Given that the object is updated, our sync loop will be triggered again and given the new status on the object will reach the add code because `cloudPrivateIPConfig.Status.Node == ""`